### PR TITLE
DropdownMenu

### LINKED ISSUE #1070: DropdownMenu
Requires #1064 and https://github.com/primer/components/issues/1056
Pre-requisite for https://github.com/primer/components/issues/1071

## Description
DropdownMenus are used to select a value from a finite, non-user-provided list of values presented in a Popover, which is propagated to the DropdownMenu’s Source.

DropdownMenus may be positioned **above** or **beside** their Source.

While type-to-select may be supported, **DropdownMenus never contain a search field**. If a search field is required, SelectPanel is a more-suitable component.

**List options according to how frequent they're used**. Don’t make users scan through a long list of disorganized options. To help users focus on what's relevant, place the options used most often at the top.

**Menus can use a Bottom sheet mobile variant** that provides room for additional items, longer descriptions, and iconography.

## Resources
Figma:
Prototype:

## Screenshots
dropdownmenu-beside-source
dropdownmenu-above-source

#### Behaviors

A DropdownMenu component renders an ActionList (`role="listbox"`) containing DropdownMenuListItems (`role="option"`) into a Popover.

note: Does the correct `role` (and related attributes, e.g. `aria-expanded`, `aria-owns`, `aria-haspopup`, `aria-autocomplete`, `aria-controls`, `aria-activedescendant`) for a DropdownMenu and its Source depend on how the DropdownMenu is positioned (above or beside) relative to its Source? I think not, because the Source is never _inside_ a DropdownMenu.

#### Behaviors

A DropdownMenuListItem component could be built on the ActionListItem component, to reduce props API surface area (i.e. so library consumers don’t need to use ActionListItem directly).

ActionMenuListItems should have the [option](https://w3c.github.io/aria/#option) role.

#### Props

name: `type` (note: as noted in [ActionListItem](https://github.com/primer/components/issues/1056), `variant` would be consistent with Primer’s existing prop naming)
type: `"singleSelection" | "multiSelection"`
default: `"singleSelection"`
description: Defines the list item type, affecting the style of the list item checkbox. (Passed through to ActionListItem.)

-----

name: `leadingVisual`
type: `React.ReactElement`
description: An optional visual to accompany the list item. It supports octicons, avatars, or other custom elements. (Passed through to ActionListItem.)

-----

name: `text`
type: `string`
description: The primary text in the list item. Use sentence case. (Passed through to ActionListItem.)

-----

name: `description`
type: `string?`
description: A quick description of the list item. (Passed through to ActionListItem.)

-----

name: `descriptionType`
type: `"inline" | "block"`
description: Defines how the description text will be placed in the list item. An inline description appears adjacent to the text prop, while block places it below in a new line. (Passed through to ActionListItem.)

### PR DESCRIPTION
**Pull request by smockle**
This is a working branch composed of `main`, `overlay` and `action-list`.

Fixes #1070 

### Preview
dropdown

### CODE REVIEW

File: src/ActionList/Item.tsx [hunk 1]
```diff
@@ -98,12 +104,19 @@ export function Item({
...
   variant = 'default',
   ...props
-}: Partial<ItemProps>): JSX.Element {
+}: Partial<ItemProps> & {item: ItemInput}): JSX.Element {
   return (
-    <StyledItem variant={variant} {...props}>
+    <StyledItem
+      tabIndex={-1}
+      variant={variant}
+      {...(!!selected === selected && {'aria-selected': selected})}
```
Comment (colebemis) [inline src/ActionList/Item.tsx:116]: What is `!!selected === selected` checking for?

Also, I tend to prefer ternary's in cases like this because I think it makes it easier to understand what the expression will evaluate too (e.g. `selected ? {...} : null` vs `selected && {...}`)

File: src/DropdownMenu/DropdownButton.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import {ChevronDownIcon} from '@primer/octicons-react'
+import Button, {ButtonProps} from '../Button/Button'
+
+export type DropdownButtonProps = ButtonProps
+
+export const DropdownButton = React.forwardRef<HTMLElement, React.PropsWithChildren<DropdownButtonProps>>(
+  ({children, ...props}: React.PropsWithChildren<DropdownButtonProps>, ref): JSX.Element => (
+    <Button ref={ref} {...props}>
+      {children}
+      <ChevronDownIcon />
```
Comment (colebemis) [inline src/DropdownMenu/DropdownButton.tsx:11]: We try to avoid using the `chevron` icon for dropdown menus ([link](https://primer.style/design/ui-patterns/progressive-disclosure#progressive-disclosure-ui-patterns)). Let's use the `TriangleDownIcon` instead. Triangle-down is what we currently use for the dropdowns on dotcom. 

I also think adding a bit of space between the text and the icon would look a litter better.

```suggestion
      <StyledOcticon icon={TriangleDownIcon} ml={1} />
```

File: docs/content/DropdownMenu.mdx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+function DemoComponent() {
+  const items = React.useMemo(() => [{text: '🔵 Cyan', id: 5}, {text: '🔴 Magenta'}, {text: '🟡 Yellow'}], [])
+  const [selectedItem, setSelectedItem] = React.useState()
+
+  return (
+    <DropdownMenu
+      renderAnchor={({children, 'aria-labelledby': ariaLabelledBy, ...anchorProps}) => (
+        <DropdownButton aria-labelledby={`favorite-color-label ${ariaLabelledBy}`} {...anchorProps}>
+          {children}
+        </DropdownButton>
```
Comment (colebemis) [inline docs/content/DropdownMenu.mdx:19]: I believe consumers will have access to the dropdown button component as `DropdownMenu.Button` not `DropdownButton`, right?

```suggestion
        <DropdownMenu.Button aria-labelledby={`favorite-color-label ${ariaLabelledBy}`} {...anchorProps}>
          {children}
        </DropdownMenu.Button>
```

File: src/DropdownMenu/DropdownMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+   * `selectedItem`, `undefined` will be passed.
+   */
+  setSelectedItem?: (item?: ItemInput) => unknown
+}
+
+/**
+ * A `DropdownMenu` provides an anchor (button by default) that will open a floating menu of selectable items.  The menu can be
+ * opened and navigated using keyboard or mouse.  When an item is selected, the menu will close and the anchor contents will be updated
+ * with the selection.
+ */
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:41]: I love helpful comments ❤️ This will be nice for our TypeScript users!

File: src/DropdownMenu/DropdownMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  /**
+   * An `ItemProps` item from the list of `items` which is currently selected.  This item will receive a checkmark next to it in the menu.
+   */
+  selectedItem?: ItemInput
+
+  /**
+   * A callback which receives the selected item or `undefined` when an item is activated in the menu.  If the activated item is the same as the current
+   * `selectedItem`, `undefined` will be passed.
+   */
+  setSelectedItem?: (item?: ItemInput) => unknown
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:34]: Could we call this `onChange` to be more consistent with other components in React like `<input>` and `<select>`?

```suggestion
  onChange?: (item?: ItemInput) => unknown
```

File: src/DropdownMenu/DropdownMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      {renderAnchor({
+        ref: anchorRef,
+        id: anchorId,
+        'aria-labelledby': anchorId,
+        'aria-haspopup': 'listbox',
+        children: selectedItem?.text ?? placeholder,
+        onClick: onAnchorClick,
+        onKeyDown: onAnchorKeyDown
+      })}
+      {open && (
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:114]: I tend to prefer ternaries over `&&` for rendering JSX. Here's a good blog post explaining why: https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx.

File: src/DropdownMenu/DropdownMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+              const itemActivated = () => {
+                setSelectedItem?.(item === selectedItem ? undefined : item)
+                onDismiss()
+              }
+
+              return renderItem({
+                ...itemProps,
+                item,
+                role: 'option',
+                selected: item === selectedItem,
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:136]: Do you think we might need a more robust equality check of objects? (for example, lodash's [`isEqual`](https://lodash.com/docs/4.17.15#isEqual) function)

File: src/DropdownMenu/index.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export {DropdownMenu} from './DropdownMenu'
+export type {DropdownMenuProps} from './DropdownMenu'
+export {DropdownButton} from './DropdownButton'
```
Comment (colebemis) [inline src/DropdownMenu/index.ts:3]: Can we namespace `DropdownButton` under `DropdownMenu` so consumers can use it like this: `DropdownMenu.Button`?

File: src/__tests__/DropdownMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,119 @@
+import {cleanup, render as HTMLRender, act, fireEvent} from '@testing-library/react'
+import 'babel-polyfill'
+import {axe, toHaveNoViolations} from 'jest-axe'
+import React from 'react'
+import theme from '../theme'
+import {DropdownMenu, DropdownButton} from '../DropdownMenu'
+import {COMMON} from '../constants'
+import {behavesAsComponent, checkExports} from '../utils/testing'
+import {ThemeProvider} from 'styled-components'
+import {BaseStyles} from '..'
```
Comment (colebemis) [inline src/__tests__/DropdownMenu.tsx:10]: We should use our custom ThemeProvider instead of the styled-components one.
```suggestion
import {BaseStyles, ThemeProvider} from '..'
```

File: src/behaviors/focusTrap.ts [hunk 10]
```diff
@@ -83,6 +84,7 @@ export function focusTrap(
         } else {
           const toFocus = initialFocus && container.contains(initialFocus) ? initialFocus : getFocusableChild(container)
           if (toFocus) {
+            prepareForFocusWithoutMouse()
```
Comment (colebemis) [inline src/behaviors/focusTrap.ts:87]: Is this going to be affected by https://github.com/primer/components/pull/1157?

File: src/index.ts [hunk 11]
```diff
@@ -78,6 +78,8 @@ export type {
   DropdownItemProps,
   DropdownMenuProps
 } from './Dropdown'
+export {DropdownButton} from './DropdownMenu/DropdownButton'
+export {DropdownMenu} from './DropdownMenu/DropdownMenu'
```
Comment (colebemis) [inline src/index.ts:82]:
```suggestion
export {DropdownMenu} from './DropdownMenu'
export type {DropdownMenuProps} from './DropdownMenu'
```

**Review by colebemis — COMMENTED** [review]
Amazing work! I'm really inspired by the amount of care y'all have put into this component ❤️ 

I left a few minor comments and questions.

File: src/DropdownMenu/index.ts (hunk 8 above)
Comment (colebemis) [inline src/DropdownMenu/index.ts:3]: Should we export `DropdownMenuButton` props as well?

File: src/ActionList/Item.tsx (hunk 1 above)
Comment (smockle) [reply src/ActionList/Item.tsx:116]: > What is `!!selected === selected` checking for?

That line is attempting to add the `aria-selected` attribute if and only if `selected` is `true` or `false`, not `undefined`. It could be rewritten as `{...(typeof selected === "boolean" && {'aria-selected': selected})}`, but since React omits attributes with `undefined` values, `aria-selected={selected}` is a readable equivalent.

> Also, I tend to prefer ternary's in cases like this because I think it makes it easier to understand what the expression will evaluate too (e.g. `selected ? {...} : null` vs `selected && {...}`)

It’s unnecessary if we use `aria-selected={selected}` as noted above, but I can accept this as a general guideline. It’s more explicit, and perhaps that will make PRC’s code more readable as a result, which is a good goal. Out of curiosity, are you aware of a lint run we could use for this (would we want a lint rule?)

Comment (smockle) [reply src/ActionList/Item.tsx]: > `aria-selected={selected}` is a readable equivalent.

Replaced in https://github.com/primer/components/pull/1135/commits/1fa39d4414a0bc317bb14e7dabdfd28f9a260224!

File: src/DropdownMenu/DropdownButton.tsx (hunk 2 above)
Comment (smockle) [reply src/DropdownMenu/DropdownButton.tsx]: Sounds good! Thanks for giving context and for providing a suggested alternative, @colebemis ❤️

Fixed in https://github.com/primer/components/pull/1135/commits/800a73878143ec3ea61080120ddf0585eb139ebf!

File: docs/content/DropdownMenu.mdx (hunk 3 above)
Comment (smockle) [reply docs/content/DropdownMenu.mdx:19]: `DropdownButton` could be used with both `DropdownMenu` and `ActionMenu`, right? If so, should it have a name/namespace that doesn’t match one of those exactly (i.e. that is more general)?

File: src/DropdownMenu/DropdownMenu.tsx (hunk 6 above)
Comment (smockle) [reply src/DropdownMenu/DropdownMenu.tsx:114]: Related, from @smockle in https://github.com/primer/components/pull/1135#discussion_r614391663:
> I can accept [preferring ternaries over `&&`] as a general guideline. Ternaries are more explicit, and perhaps that will make PRC’s code more readable as a result, which is a good goal. Out of curiosity, [is anyone] aware of a lint rule we could use for this (would we want a lint rule?)

File: src/DropdownMenu/index.ts (hunk 8 above)
Comment (smockle) [reply src/DropdownMenu/index.ts:3]: Related, from @smockle in https://github.com/primer/components/pull/1135#discussion_r614395895:
> `DropdownButton` could be used with both `DropdownMenu` and `ActionMenu`, right? If so, should it have a name/namespace that doesn’t match one of those exactly (i.e. that is more general)?

File: docs/content/DropdownMenu.mdx (hunk 3 above)
Comment (colebemis) [reply docs/content/DropdownMenu.mdx:19]: Ah, I see. I didn't realize it was used with ActionMenu as well. In that case, I think leaving it as `DropdownButton` is fine 👍

File: src/DropdownMenu/DropdownMenu.tsx (hunk 6 above)
Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx]: Fantastic article with real world reasons why this is important! Fixed in https://github.com/primer/components/pull/1135/commits/710123afbb5548be9405e42d8465b80894764e43

File: src/DropdownMenu/DropdownMenu.tsx (hunk 7 above)
Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx:136]: This was an intentional design decision to avoid any complications with duplicate or similar items.  We talked about comparing `text`, but that would be very easy to have duplicates.  We could also require an `id` prop on each item, but that's very restrictive.  In the end, the only way to ensure we have the _exact_ item selected is to do an object equality check.  Happy to hear other ideas if you have them.

File: src/behaviors/focusTrap.ts (hunk 10 above)
Comment (dgreif) [reply src/behaviors/focusTrap.ts:87]: This was a hacked fix before #1157.  I'll get master merged in and remove this.

File: src/index.ts (hunk 11 above)
Comment (dgreif) [reply src/index.ts:82]: I think this is a good change, however there is a conflict with the types exported above: https://github.com/primer/components/blob/0b5678143c2c4be93cecc955bf6547b775163595/src/index.ts#L74-L80

To avoid the conflict, I'd say we just export the new menu and button without the prop types for now

File: src/__tests__/DropdownMenu.tsx (hunk 9 above)
Comment (dgreif) [reply src/__tests__/DropdownMenu.tsx]: Good catch.  Fixed in https://github.com/primer/components/pull/1135/commits/33863d8276f47d2addfd2aef0d2554569edb4a38.  Found one other use as well

File: src/DropdownMenu/index.ts (hunk 8 above)
Comment (dgreif) [reply src/DropdownMenu/index.ts:3]: Sounds like a good idea.  Done in https://github.com/primer/components/pull/1135/commits/edf01d5694dbb3df8a7336a156ba45ec241836e9

File: src/DropdownMenu/DropdownMenu.tsx (hunk 5 above)
Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx]: Changed in https://github.com/primer/components/pull/1135/commits/7e74ab11283dd3f7de2668a5aebb777190ef0df4

File: src/index.ts (hunk 11 above)
Comment (colebemis) [reply src/index.ts]: Works for me 👍 Maybe we should leave a comment about why they're not exported

File: src/DropdownMenu/DropdownMenu.tsx (hunk 7 above)
Comment (colebemis) [reply src/DropdownMenu/DropdownMenu.tsx:136]: That's fine with me. Just wanted to call attention to it in case it was unintentional :)

File: src/DropdownMenu/index.ts (hunk 8 above)
Comment (smockle) [reply src/DropdownMenu/index.ts:3]: Reply from @colebemis in https://github.com/primer/components/pull/1135#discussion_r614405677:
> Ah, I see. I didn't realize it was used with `ActionMenu` as well. In that case, I think leaving it as `DropdownButton` is fine 👍

File: src/DropdownMenu/DropdownMenu.tsx [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+            setFocusType('list')
+            setOpen(true)
+            event.preventDefault()
+          } else if (event.key === ' ' || event.key === 'Enter') {
+            setFocusType('anchor')
+            setOpen(true)
+            event.preventDefault()
+          }
+        } else if (focusType === 'anchor') {
+          if (['ArrowDown', 'ArrowUp', 'Tab', 'Enter'].indexOf(event.key) !== -1) {
```
Comment (T-Hugs) [inline src/DropdownMenu/DropdownMenu.tsx:76]:
```suggestion
          if (['ArrowDown', 'ArrowUp', 'Tab', 'Enter'].includes(event.key)) {
```

File: src/DropdownMenu/DropdownMenu.tsx [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+      {renderAnchor({
+        ref: anchorRef,
+        id: anchorId,
+        'aria-labelledby': anchorId,
+        'aria-haspopup': 'listbox',
+        children: selectedItem?.text ?? placeholder,
+        onClick: onAnchorClick,
+        onKeyDown: onAnchorKeyDown
+      })}
+      {open ? (
```
Comment (T-Hugs) [inline src/DropdownMenu/DropdownMenu.tsx:114]: Instead of `{foo ? <SomeJSX /> : null}` it's more idiomatic to use `{foo && <SomeJSX />}`

File: src/DropdownMenu/DropdownMenu.tsx [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+  useFocusZone({containerRef: overlayRef, disabled: !open || focusType !== 'list' || !position})
+  useFocusTrap({containerRef: overlayRef, disabled: !open || focusType !== 'list' || !position})
+
+  return (
+    <>
+      {renderAnchor({
+        ref: anchorRef,
+        id: anchorId,
+        'aria-labelledby': anchorId,
```
Comment (T-Hugs) [inline src/DropdownMenu/DropdownMenu.tsx:108]: Why is the anchor's `aria-labelledby` referencing itself? Is this intentional?

File: src/DropdownMenu/DropdownMenu.tsx [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+   */
+  onChange?: (item?: ItemInput) => unknown
+}
+
+/**
+ * A `DropdownMenu` provides an anchor (button by default) that will open a floating menu of selectable items.  The menu can be
+ * opened and navigated using keyboard or mouse.  When an item is selected, the menu will close and the `onChange` callback will be called.
+ * If the default anchor button is used, the anchor contents will be updated with the selection.
+ */
+export function DropdownMenu({
```
Comment (T-Hugs) [inline src/DropdownMenu/DropdownMenu.tsx:42]: There are quite a few anonymous functions defined inline in this function component (e.g. L126, L137, L141). Do these need to be memoized with `useCallback`?

File: src/DropdownMenu/DropdownMenu.tsx (hunk 13 above)
Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx:114]: See https://github.com/primer/components/pull/1135#discussion_r614371832.  @colebemis linked a great article explaining why ternary is much safer to use in these situations.

File: src/__tests__/Overlay.tsx [hunk 16]
```diff
@@ -27,10 +27,8 @@ const TestComponent = ({initialFocus, callback}: TestComponentSettings) => {
         open overlay
       </Button>
       <Button>outside</Button>
-      {isOpen && (
+      {isOpen ? (
```
Comment (T-Hugs) [inline src/__tests__/Overlay.tsx:30]: why this change?

File: src/hooks/useFocusTrap.ts [hunk 17]
```diff
@@ -28,25 +28,30 @@ interface FocusTrapHookSettings {
  * @param settings {FocusTrapHookSettings}
  */
 export function useFocusTrap(
-  settings?: FocusTrapHookSettings
+  settings?: FocusTrapHookSettings,
+  dependencies?: React.DependencyList
```
Comment (T-Hugs) [inline src/hooks/useFocusTrap.ts:32]: Like the `useAnchoredPosition` change above, maybe do the same here?

```suggestion
  dependencies: React.DependencyList = []
```

File: src/hooks/useFocusZone.ts [hunk 18]
```diff
@@ -36,26 +36,29 @@ export function useFocusZone(
...
+            activeDescendantControl: activeDescendantControlRef.current ?? undefined
+          }
+          abortController.current = focusZone(containerRef.current, vanillaSettings)
+          return () => {
+            abortController.current?.abort()
+          }
+        } else {
           abortController.current?.abort()
         }
-      } else {
```
Comment (T-Hugs) [inline src/hooks/useFocusZone.ts:53]: Why was this removed?

File: src/DropdownMenu/DropdownMenu.tsx [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+  renderItem = Item,
+  placeholder,
+  selectedItem,
+  onChange,
+  ...listProps
+}: DropdownMenuProps): JSX.Element {
+  const anchorRef = useRef<HTMLElement>(null)
+  const [overlayRef, updateOverlayRef] = useRenderForcingRef<HTMLDivElement>()
+
+  const anchorId = `dropdownMenuAnchor-${randomId()}`
```
Comment (T-Hugs) [inline src/DropdownMenu/DropdownMenu.tsx:53]: Why not use `uniqueId`?

**Review by T-Hugs — CHANGES_REQUESTED** [review]
Mostly minor comments. Overall it looks great; can't wait to get this merged! 🎸

File: src/DropdownMenu/DropdownMenu.tsx (hunk 6 above)
Comment (T-Hugs) [reply src/DropdownMenu/DropdownMenu.tsx]: Fair enough. I want to propose a new operator, `!!`, which is the counterpart to `??`, basically "If LHS is null or undefined, evaluate to LHS, otherwise evaluate to RHS".

File: src/__tests__/Overlay.tsx (hunk 16 above)
Comment (dgreif) [reply src/__tests__/Overlay.tsx:30]: I changed all the && to ternary based on https://github.com/primer/components/pull/1135#discussion_r614371832.  Could definitely have done it in a separate PR, but there were only a few

File: src/DropdownMenu/DropdownMenu.tsx (hunk 13 above)
Comment (T-Hugs) [reply src/DropdownMenu/DropdownMenu.tsx:114]: Eh, kind of a weak argument if you ask me, but fair enough!

Comment (T-Hugs) [reply src/DropdownMenu/DropdownMenu.tsx:114]: The tradeoff here is a ternary expression that potentially spans dozens of lines, which can get really hard to read.

Comment (T-Hugs) [reply src/DropdownMenu/DropdownMenu.tsx:114]: New lint rule proposal: No `&&` operators within a JSX expression such that the left side of the operator is a number or a string type :)

File: src/hooks/useFocusZone.ts (hunk 18 above)
Comment (dgreif) [reply src/hooks/useFocusZone.ts:53]: Looks like this is just a whitespace change.  Git diff ended up weird because of it.  The only piece that actually changed here was the deps array

Comment (T-Hugs) [reply src/hooks/useFocusZone.ts:53]: Whoops, you're right!

File: src/DropdownMenu/DropdownMenu.tsx (hunk 19 above)
Comment (VanAnderson) [reply src/DropdownMenu/DropdownMenu.tsx:53]: Ahh, this was my doing here - I'm not familiar with uniqueId, is that something that will natively accomplish the hand rolled method I did here?

File: docs/content/DropdownMenu.mdx [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+| :--------------- | :-------------------------------------------- | :---------------: | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| items            | `ItemProps[]`                                 |    `undefined`    | Required. A list of item objects to display in the menu                                                                                                                                        |
+| selectedItem     | `ItemInput`                                   |    `undefined`    | An `ItemProps` item from the list of `items` which is currently selected. This item will receive a checkmark next to it in the menu.                                                           |
+| setSelectedItem? | (item?: ItemInput) => unknown                 |    `undefined`    | A callback which receives the selected item or `undefined` when an item is activated in the menu. If the activated item is the same as the current `selectedItem`, `undefined` will be passed. |
```
Comment (colebemis) [inline docs/content/DropdownMenu.mdx:38]:
```suggestion
| onChange? | (item?: ItemInput) => unknown                 |    `undefined`    | A callback which receives the selected item or `undefined` when an item is activated in the menu. If the activated item is the same as the current `selectedItem`, `undefined` will be passed. |
```

File: src/DropdownMenu/DropdownMenu.tsx [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+          onEscape={onDismiss}
+          ref={updateOverlayRef}
+          {...position}
+        >
+          <List
+            {...listProps}
+            role="listbox"
+            renderItem={({onClick, onKeyDown, item, ...itemProps}) => {
+              const itemActivated = () => {
+                onChange?.(item === selectedItem ? undefined : item)
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:128]: Maybe we shouldn't call `onChange` if the selectedItem didn't change?

File: src/DropdownMenu/DropdownMenu.tsx [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+                onDismiss()
+              }
+
+              return renderItem({
+                ...itemProps,
+                item,
+                role: 'option',
+                selected: item === selectedItem,
+                onClick: event => {
+                  itemActivated()
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:138]: This function name was a little confusing to me at first. Maybe `handleSelection()` would be more clear?

File: src/DropdownMenu/DropdownMenu.tsx (hunk 21 above)
Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx:128]: Great catch!

Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx:128]: The current implementation is to call `onChange` with `undefined` if the currently selected item is activated.  This allows the item to be "deselected" if the consuming apps wants to support that behavior.  If they don't, they can ignore `undefined` in the callback.

File: src/DropdownMenu/DropdownMenu.tsx [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+          initialFocusRef={anchorRef}
+          returnFocusRef={anchorRef}
+          onClickOutside={onDismiss}
+          onEscape={onDismiss}
+          ref={updateOverlayRef}
+          {...position}
+        >
+          <List
+            {...listProps}
+            role="listbox"
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:125]: I love the attention to accessibility in this component ❤️

File: src/DropdownMenu/DropdownMenu.tsx [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+                  itemActivated()
+                  onClick && onClick(event)
+                },
+                onKeyDown: event => {
+                  if (!event.defaultPrevented && [' ', 'Enter'].includes(event.key)) {
+                    itemActivated()
+                    // prevent "Enter" event from becoming a click on the anchor as overlay closes
+                    event.preventDefault()
+                  }
+                  onKeyDown && onKeyDown(event)
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:147]: I believe we can use the `?.` operator to clean this up a bit:
```suggestion
                  onClick?.(event)
                },
                onKeyDown: event => {
                  if (!event.defaultPrevented && [' ', 'Enter'].includes(event.key)) {
                    itemActivated()
                    // prevent "Enter" event from becoming a click on the anchor as overlay closes
                    event.preventDefault()
                  }
                  onKeyDown?.(event)
```

File: src/DropdownMenu/DropdownButton.tsx [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+export const DropdownButton = React.forwardRef<HTMLElement, React.PropsWithChildren<DropdownButtonProps>>(
+  ({children, ...props}: React.PropsWithChildren<DropdownButtonProps>, ref): JSX.Element => (
+    <Button ref={ref} {...props}>
+      {children}
+      <StyledOcticon icon={TriangleDownIcon} ml={1} />
+    </Button>
+  )
+)
+
+DropdownButton.displayName = 'DropdownMenu.Button'
```
Comment (colebemis) [inline src/DropdownMenu/DropdownButton.tsx:17]: If we're not namespacing `DropdownButton` under `DropdownMenu`, we probably don't need to change the display name.

```suggestion
```

File: src/DropdownMenu/DropdownMenu.tsx [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+
+  const onAnchorKeyDown = useCallback(
+    (event: React.KeyboardEvent<HTMLElement>) => {
+      if (!event.defaultPrevented) {
+        if (!open) {
+          if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+            setFocusType('list')
+            setOpen(true)
+            event.preventDefault()
+          } else if (event.key === ' ' || event.key === 'Enter') {
```
Comment (colebemis) [inline src/DropdownMenu/DropdownMenu.tsx:70]: [Minor] We could clean these up to be consistent with the if statement on line 76. I don't feel too strongly about this though.
```suggestion
          if (['ArrowDown', 'ArrowUp'].includes(event.key)) {
            setFocusType('list')
            setOpen(true)
            event.preventDefault()
          } else if ([' ', 'Enter'].includes(event.key)) {
```

File: src/DropdownMenu/DropdownMenu.tsx (hunk 21 above)
Comment (colebemis) [reply src/DropdownMenu/DropdownMenu.tsx:128]: Ah, I see. That makes sense.

File: src/DropdownMenu/DropdownMenu.tsx (hunk 19 above)
Comment (T-Hugs) [reply src/DropdownMenu/DropdownMenu.tsx:53]: @VanAnderson It's not native, but`uniqueId` is defined [here](https://github.com/primer/components/blob/main/src/utils/uniqueId.ts). Instead of a random ID, it produces an ID with a monotonically increasing value to guarantee uniqueness.

Comment (VanAnderson) [reply src/DropdownMenu/DropdownMenu.tsx:53]: Gotcha, @dgreif I'll update this in this PR and in `ActionMenu` 👍

Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx]: Refactored in https://github.com/primer/components/pull/1135/commits/8eabeeed67e13b5a8a1d3d1ad2ac33f14c962c9b

File: src/DropdownMenu/DropdownMenu.tsx (hunk 14 above)
Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx:108]: Looks like it is accurate to label the button with it's own id based on https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html.  If there is an outside label, it should be mixed in first as is done in the story: https://github.com/primer/components/pull/1135/files#diff-14de3214a329b0c98400ce1c0a330f58e8cc55fcb14981be4b1e0951ee556cc2R40

File: src/DropdownMenu/DropdownMenu.tsx (hunk 15 above)
Comment (dgreif) [reply src/DropdownMenu/DropdownMenu.tsx:42]: Done in https://github.com/primer/components/pull/1135/commits/90546176e9e1285fb8b3fc3f69bb19a1d9d69d00

File: src/DropdownMenu/DropdownMenu.tsx (hunk 14 above)
Comment (T-Hugs) [reply src/DropdownMenu/DropdownMenu.tsx:137]: TIL!

**Review by T-Hugs — APPROVED** [review]
👌 💯 🚀 🎸

**Review by colebemis — APPROVED** [review]
🎉 🚢
